File: yeongmok/7주차(Binary Search)/입국심사.java
import java.util.*;

class Solution {
    Map<Integer, Integer> map = new HashMap<>();
    
    public long solution(int n, int[] times) {
        Arrays.sort(times);
        
        Long T=(long) times[0]*n;
        Long answer = binary(n,T,times);
        
        return answer;
    }
    
    Long binary(int n,Long T, int[] times){
        Long low=0L;
        Long high=T;
        
        while(low<high){
            Long mid = (low+high)/2;
            Long p=0L;
            
            for(int i=0;i<times.length;i++){
                p+=mid/times[i];
            }
            if(p>=n){
                high=mid;
            }
            else{
                low = mid+1;
            }
        }
        return low;
    }
}

File: hayan/7주차_이분탐색/입국심사.py
def solution(n, times):    
    l =1
    r =n*max(times)
    answer = r
    
    while l<=r:
        mid = (l+r)//2
        n_sum = 0
        
        for i in range(len(times)):
            n_sum += mid//times[i]
            
        if n_sum >= n: 
            answer = mid
            r = mid-1
        else:
            l = mid+1
            
    return answer